indexer/sql/src/non_blocking_indexer.rs

```diff
@@ -0,0 +1,343 @@
+    pub txs: Vec<(Tx, TxOutcome)>,
+}
+
+impl BlockToIndex {
+    /// Takes care of inserting the data in the database
+    pub async fn save(self, db: &DatabaseTransaction) -> error::Result<()> {
+        #[cfg(feature = "tracing")]
+        tracing::info!(block_height = self.block_info.height, "Indexing block");
+
+        let mut models = Models::build(&self.block_info)?;
+        for tx in self.txs.iter() {
+            models.push(&tx.0, &tx.1)?;
+        }
+
+        // TODO: if the process was to crash in the middle and restarted, we could try to
+        // reinsert existing data. We should use `on_conflict()` to avoid this, return the
+        // existing block and change `block_id` when/if we added foreign keys
+        models.block.insert(db).await?;
+        entity::transactions::Entity::insert_many(models.transactions)
+            .exec(db)
+            .await?;
+        entity::messages::Entity::insert_many(models.messages)
+            .exec(db)
+            .await?;
+        entity::events::Entity::insert_many(models.events)
+            .exec(db)
+            .await?;
+        Ok(())
+    }
+}
+
+impl NonBlockingIndexer {
+    fn find_or_create<F, R>(&self, block: &BlockInfo, action: F) -> error::Result<R>
+    where
+        F: FnOnce(&mut BlockToIndex) -> error::Result<R>,
+    {
+        let mut blocks = self.blocks.lock().expect("Can't lock blocks");
+        let block_to_index = blocks.entry(block.height).or_insert(BlockToIndex {
+            block_info: *block,
+            txs: vec![],
+        });
+        action(block_to_index)
+    }
+
+    fn find_or_fail(&self, block_height: &u64) -> error::Result<BlockToIndex> {
+        let blocks = self.blocks.lock().expect("Can't lock blocks");
+
+        let block_to_index = match blocks.get(block_height) {
+            Some(block_to_index) => block_to_index,
+            None => {
+                bail!("Block {} not found", block_height);
+            },
+        };
+        Ok(block_to_index.clone())
+    }
+
+    fn remove_or_fail(
+        blocks: Arc<Mutex<HashMap<u64, BlockToIndex>>>,
+        block_height: &u64,
+    ) -> error::Result<BlockToIndex> {
+        let mut blocks = blocks.lock().expect("Can't lock blocks");
+        let block_to_index = match blocks.remove_entry(block_height) {
+            Some(block) => block,
+            None => {
+                bail!("Block {} not found", block_height);
+            },
+        };
+        #[cfg(feature = "tracing")]
+        tracing::debug!(
+            block_height = block_height,
+            blocks_len = blocks.len(),
+            "remove_or_fail called"
+        );
+        Ok(block_to_index.1)
+    }
+}
+
+impl Indexer for NonBlockingIndexer {
+    type Error = crate::error::IndexerError;
+
+    fn start(&mut self) -> error::Result<()> {
+        let context = self.context.clone();
+        block_call(self.runtime.as_ref(), &self.handle, async move {
+            context.migrate_db().await
+        })?;
```

Comment:
There is no need to clone `self.context`. `Context.migrate_db()` take `&self` as param, so you can remove `move` keyworks and use `self.context.migrate_db()`.
```suggestion
        block_call(self.runtime.as_ref(), &self.handle, async {
            self.context.migrate_db().await
        })?;
```